libs/community/tests/unit_tests/vectorstores/test_sqlserver.py

```diff
@@ -519,13 +471,34 @@ def test_that_case_sensitivity_does_not_affect_distance_strategy(
     conn.close()
 
 
+def test_that_rows_with_duplicate_custom_id_cannot_be_entered(
+    store: SQLServer_VectorStore,
+    texts: List[str],
+) -> None:
+    """Test that if a row is specified with existing ID in the table,
+    `add_texts` fails."""
+    metadatas = [
+        {"id": 1, "summary": "Good Quality Dog Food"},
+        {"id": 2, "summary": "Nasty No flavor"},
+        {"id": 2, "summary": "stale product"},
+        {"id": 4, "summary": "Great value and convenient ramen"},
+        {"id": 5, "summary": "Great for the kids!"},
+    ]
+    with pytest.raises(Exception):
+        store.add_texts(texts, metadatas)
+
+
 # We need to mock this so that actual connection is not attempted
 # after mocking _provide_token.
 @mock.patch("sqlalchemy.dialects.mssql.dialect.initialize")
 @mock.patch(
     "langchain_community.vectorstores.sqlserver.SQLServer_VectorStore._provide_token"
 )
+@mock.patch(
+    "langchain_community.vectorstores.sqlserver.SQLServer_VectorStore._prepare_json_data_type"
```

Comment:
Yes, they are related.

For entra id authentication, recall that we need to call `_provide_token` to get auth token that then enables us to connect to the DB, however, we are now mocking the call to `_provide_token` to be able to check if it's called, this means we are not actually going through with the `_provide_token` call to get token and subsequently, every attempt to connect will fail.

`prep_json_data_type` attempts a connection to the DB in order to check for the data type, but we're already mocking an important part of that workflow and it's  bound to fail since we're not providing a valid token. This is why we mock `_prepare_json_data_type` too.

Reply:
Sure, these tests would be better suited for a test suite where the entirety of the test is based on Mocks. When we do create our mock tests, we can move this into those tests and modify as needed.